### ISSUE TITLE
Install fail (another) on python 3.6 windows 10

### ISSUE THREAD

**Issue by Taverius** [saghul/pyuv#238]
```
> pip3 install pyuv
Collecting pyuv
  Using cached pyuv-1.3.0.tar.gz
Installing collected packages: pyuv
  Running setup.py install for pyuv ... error
    Complete output from command c:\python36\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\Taverius\\AppData\\Local\\Temp\\pip-build-nocsm8qv\\pyuv\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\Taverius\AppData\Local\Temp\pip-v0oqpbne-record\install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build\lib.win-amd64-3.6
    creating build\lib.win-amd64-3.6\pyuv
    copying pyuv\_version.py -> build\lib.win-amd64-3.6\pyuv
    copying pyuv\__init__.py -> build\lib.win-amd64-3.6\pyuv
    running build_ext
    libuv needs to be compiled.
    Building libuv...
    Using "C:\Python27\python.exe" to build libuv...
    building 'pyuv._cpyuv' extension
    creating build\temp.win-amd64-3.6
    creating build\temp.win-amd64-3.6\Release
    creating build\temp.win-amd64-3.6\Release\src
    C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DWIN32=1 -Ic:\python36\include -Ic:\python36\include -Ideps\libuv\include "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\winrt" /Tcsrc/pyuv.c /Fobuild\temp.win-amd64-3.6\Release\src/pyuv.obj
    pyuv.c
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\loop.c(114): warning C4311: 'type cast': pointer truncation from 'uv_work_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\loop.c(136): warning C4311: 'type cast': pointer truncation from 'uv_work_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\async.c(10): warning C4311: 'type cast': pointer truncation from 'uv_async_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\timer.c(10): warning C4311: 'type cast': pointer truncation from 'uv_timer_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\prepare.c(10): warning C4311: 'type cast': pointer truncation from 'uv_prepare_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\idle.c(10): warning C4311: 'type cast': pointer truncation from 'uv_idle_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\check.c(10): warning C4311: 'type cast': pointer truncation from 'uv_check_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\signal.c(11): warning C4311: 'type cast': pointer truncation from 'uv_signal_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\stream.c(28): warning C4311: 'type cast': pointer truncation from 'uv_shutdown_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\stream.c(113): warning C4311: 'type cast': pointer truncation from 'uv_write_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\stream.c(269): warning C4244: 'function': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\stream.c(312): warning C4244: 'function': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\stream.c(377): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\stream.c(387): warning C4244: '=': conversion from 'Py_ssize_t' to 'ULONG', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\stream.c(399): warning C4244: 'function': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\stream.c(401): warning C4244: 'function': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\stream.c(475): warning C4311: 'type cast': pointer truncation from 'uv_os_fd_t' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\pipe.c(10): warning C4311: 'type cast': pointer truncation from 'uv_pipe_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\pipe.c(42): warning C4311: 'type cast': pointer truncation from 'uv_pipe_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\tcp.c(11): warning C4311: 'type cast': pointer truncation from 'uv_tcp_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\tcp.c(43): warning C4311: 'type cast': pointer truncation from 'uv_tcp_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\udp.c(22): warning C4311: 'type cast': pointer truncation from 'uv_udp_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\udp.c(81): warning C4311: 'type cast': pointer truncation from 'uv_udp_send_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\udp.c(82): warning C4311: 'type cast': pointer truncation from 'uv_udp_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\udp.c(230): warning C4244: 'function': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\udp.c(270): warning C4244: 'function': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\udp.c(329): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\udp.c(339): warning C4244: '=': conversion from 'Py_ssize_t' to 'ULONG', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\udp.c(345): warning C4244: 'function': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\udp.c(611): warning C4311: 'type cast': pointer truncation from 'uv_os_fd_t' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\poll.c(11): warning C4311: 'type cast': pointer truncation from 'uv_poll_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\poll.c(113): warning C4311: 'type cast': pointer truncation from 'uv_os_fd_t' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\fs.c(65): warning C4311: 'type cast': pointer truncation from 'uv_fs_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\fs.c(1058): warning C4244: 'function': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\fs.c(1582): warning C4311: 'type cast': pointer truncation from 'uv_fs_event_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\fs.c(1836): warning C4311: 'type cast': pointer truncation from 'uv_fs_poll_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\process.c(185): warning C4311: 'type cast': pointer truncation from 'uv_process_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\process.c(187): warning C4244: 'function': conversion from 'int64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\dns.c(57): warning C4311: 'type cast': pointer truncation from 'uv_getaddrinfo_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\dns.c(113): warning C4311: 'type cast': pointer truncation from 'uv_getnameinfo_t *' to 'long'
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(220): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(221): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(222): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(223): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(224): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(225): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(226): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(227): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(228): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(229): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(230): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(231): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(232): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(233): warning C4244: 'function': conversion from 'uint64_t' to 'long', possible loss of data
    c:\users\taverius\appdata\local\temp\pip-build-nocsm8qv\pyuv\src\util.c(313): warning C4311: 'type cast': pointer truncation from 'uv_poll_t *' to 'long'
    C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:c:\python36\libs /LIBPATH:c:\python36\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\LIB\amd64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.15063.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.15063.0\um\x64" advapi32.lib iphlpapi.lib psapi.lib shell32.lib user32.lib userenv.lib ws2_32.lib /EXPORT:PyInit__cpyuv build\temp.win-amd64-3.6\Release\src/pyuv.obj deps\libuv\Release\lib\libuv.lib /OUT:build\lib.win-amd64-3.6\pyuv\_cpyuv.cp36-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.6\Release\src\_cpyuv.cp36-win_amd64.lib /NODEFAULTLIB:libcmt /LTCG
    pyuv.obj : warning LNK4197: export 'PyInit__cpyuv' specified multiple times; using first specification
       Creating library build\temp.win-amd64-3.6\Release\src\_cpyuv.cp36-win_amd64.lib and object build\temp.win-amd64-3.6\Release\src\_cpyuv.cp36-win_amd64.exp
    Generating code
    Finished generating code
    LINK : fatal error LNK1158: cannot run 'rc.exe'
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\x86_amd64\\link.exe' failed with exit status 1158
```

Assuming the env was setup correctly it should be added to the paths ... `rc` *is* present in the \bin associated with the SDK `/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.15063.0\` its linking to - `C:\Program Files (x86)\Windows Kits\10\bin\10.0.15063.0\x64` so it looks like a configuration error.

Any ideas?

**Comment by schlamar** [comment]
It's working with VS2015 on [AppVeyor](https://ci.appveyor.com/project/saghul/pyuv/build/job/i97cspw992wir0bv) so it's likely a configuration issue on your side. Have you tried reinstalling the Windows 10 SDK?

**Comment by schlamar** [comment]
It's also working in a Windows 10 VM with a fresh install of VS2015 Build Tools (Windows 8 and Windows 10 SDK checked in features).

**Comment by fuzzybear3965** [comment]
It's not building for me, either. I've installed both the Windows 10 SDK build tools and the Microsoft Build Tools 2015. I'm on Windows 10 using Python 3.6.1.

```
λ pip install neovim
Collecting neovim
  Using cached neovim-0.1.13.tar.gz
Requirement already satisfied: msgpack-python>=0.4.0 in c:\python36\lib\site-packages (from neovim)
Collecting pyuv>=1.0.0 (from neovim)
  Using cached pyuv-1.3.0.tar.gz
Collecting greenlet (from neovim)
  Using cached greenlet-0.4.12-cp36-cp36m-win_amd64.whl
Installing collected packages: pyuv, greenlet, neovim
  Running setup.py install for pyuv ... error
    Complete output from command c:\python36\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\JohnRinehart\\AppData\\Local\\Temp\\pip-build-1e3_i6w6\\pyuv\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\JohnRinehart\AppData\Local\Temp\pip-errv9fsm-record\install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build\lib.win-amd64-3.6
    creating build\lib.win-amd64-3.6\pyuv
    copying pyuv\_version.py -> build\lib.win-amd64-3.6\pyuv
    copying pyuv\__init__.py -> build\lib.win-amd64-3.6\pyuv
    running build_ext
    libuv needs to be compiled.
    Building libuv...
    Using "C:\Python27\python.exe" to build libuv...
    error: Got return value 1 while executing "cmd.exe /C vcbuild.bat x64 release", output was:
    stderr:
      File "C:\Python36\\lib\site.py", line 177
        file=sys.stderr)
            ^
    SyntaxError: invalid syntax
    stdout:
    Failed to create vc project files.

    ----------------------------------------
Command "c:\python36\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\JohnRinehart\\AppData\\Local\\Temp\\pip-build-1e3_i6w6\\pyuv\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\JohnRinehart\AppData\Local\Temp\pip-errv9fsm-record\install-record.txt --single-version-externally-managed --compile" failed with error code 1 in C:\Users\JohnRinehart\AppData\Local\Temp\pip-build-1e3_i6w6\pyuv\
```

```
C:\Users\JohnRinehart
λ which pip
/c/Python36/Scripts/pip
```

**Comment by Taverius** [comment]
@schlamar Weird.

SDK 8/10 are installed.

Might it be because 2017 build tools are installed as well?

I'm not in a position where I can pull that out of my machine to test :/

**Comment by Taverius** [comment]
Just in case I pulled out the Windows 7.1 SDK & 2010 compilers - don't need those anymore - and repaired my 2015 build tools install, still no cigar.

I can make it work manually by adding the SDK bin to the PATH, but it would be best to understand why its breaking.

**Comment by schlamar** [comment]
@fuzzybear3965 There is something wrong with your Python 2.7 setup. It looks like this is importing Python 3.6 packages (C:\Python36\\lib\site.py)?! Whats the output of `C:\Python27\python.exe -c "import site; print site.getsitepackages(); import sys; print sys.path"`?

@Taverius There are a lot of Google hits regarding this issue which are unrelated to building Python extensions so I'm pretty sure this is a Microsoft issue.

**Comment by schlamar** [comment]
@fuzzybear3965 See https://stackoverflow.com/questions/30357973/python-2-7-still-accessing-python-3-4-directories. You should fix your environment variables.

**Comment by schlamar** [comment]
@Taverius Have you installed the VS 2015 build tools with Windows 10 SDK enabled as feature? I assume Windows SDK 10 installed seperately is not compatible with VS 2015 build tools. On my setup the build is using Windows 10 SDK version 10.0.10240.0 (`C:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0`).

![temp](https://cloud.githubusercontent.com/assets/238652/26626006/16dc6f6e-45f6-11e7-9215-a60c8870cd5f.png)

**Comment by schlamar** [comment]
The command 

    python -c "from distutils import _msvccompiler; print(_msvccompiler._get_vc_env('x64')['path'])" 

should give you a result with `C:\Program Files (x86)\Windows Kits\10\bin\x64` included. 

Otherwise your installation is somehow corrupt. In this case you can debug distutils/_msvccompiler.py and `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat` to check what's wrong with your setup.

**Comment by fuzzybear3965** [comment]
Hey, @schlamar, you're right; that's what was happening with me. I'll be careful to document all the steps I use in order to successfully build on Windows 10 using Python 3.6. The use of a virtual environment and the Windows 10 SDK could prove challenging for other users (and is proving challenging for me). Could you migrate the build system to using VS 2017? It's the most recently supported version of Visual Studio. It's [some work](https://www.visualstudio.com/vs/older-downloads/) to get VS 2015. Also, should this build work as long as we're installing the [Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk)? I personally don't want to install the bloat associated with VS 2015 (especially for pyuv, which is, itself, very light).

**Comment by schlamar** [comment]
> Could you migrate the build system to using VS 2017?

VS 2017 should work after #240 is done. Right now, libuv itself cannot be compiled with VS 2017.

> don't want to install the bloat associated with VS 2015

There is no need to install Visual Studio. You can use the stand alone Visual C++ 2015 Build Tools (with SDK 8.1 and SDK 10 checked as features) from http://landinghub.visualstudio.com/visual-cpp-build-tools.

> Also, should this build work as long as we're installing the Windows 10 SDK?

I'll check this in my Windows VM and report back.

**Comment by schlamar** [comment]
BTW, I hope there will be a new pyuv release very soon with Python 3.6 wheels, so there is no need to compile it on your own anymore.

**Comment by Taverius** [comment]
@schlamar its there.

> C:\Program Files (x86)\MSBuild\14.0\bin\amd64;C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64;C:\Windows\Microsoft.NET\Framework64\v4.0.30319;C:\Windows\Microsoft.NET\Framework64\;**C:\Program Files (x86)\Windows Kits\10\bin\x64**;*SNIP*

I compiles fine if I manually add the path, so at this point we're down to figuring out *why* it fails when it so clearly shouldn't

**Comment by schlamar** [comment]
>> Also, should this build work as long as we're installing the Windows 10 SDK?
>
> I'll check this in my Windows VM and report back.

I just installed Windows 10 SDK (10.0.15063.400) on top of the 2015 Build Tools and it still works.

**Comment by schlamar** [comment]
@Taverius Seems like distutils is using the x86 x64 cross compiler, so please check output of 

    python -c "from distutils import _msvccompiler; print(_msvccompiler._get_vc_env('x86_amd64')['path'])"

**Comment by fuzzybear3965** [comment]
I reviewed all of #240 and [1284 from libuv/libuv](https://github.com/libuv/libuv/pull/1284#issuecomment-305099237). I had some difficulty understanding what's going on (what the goals are and what the problems are). I understand that, because of the way that VS 2017 makes its presence known to the OS it's no longer possible to do some environment variable detection of a VS 2017 installation.

1) However, I don't understand how migrating to a new version of libuv fixes this (per #240). I also don't understand what is meant by "ditching the libuv build system". I understand that libuv is compiled using the Windows SDK upon installation of pyuv, right now.

2) But, I don't understand what is meant by "ditching the libuv build system". libuv will still need to be compiled for pyuv, right? If libuv sources (written in C) are included in pyuv (written in Python), then it still seems like some shared C library will need to be compiled in order to make libuv acccessible by pyuv.

Thanks for helping me understand what's going on. I'll report later with more tests (it make might take me a while to attempt another installation - I have a busy morning).

**Comment by Taverius** [comment]
@schlamar still there, but I figured out the issue, I think.

Because I installed VCTools 2017 *first*, and then VCTools 2015 *after*, the main part of the SDK is in, for example, in `C:\Program Files (x86)\Windows Kits\10\bin\10.0.15063.0\x86\`, not in `C:\Program Files (x86)\Windows Kits\10\bin\x86\`.

The stuff not `bin\<version>\arch` seems to be just the difference, so you can overlay it over the base version.

If I understand what M$ is trying to do here.

I could be wrong, but the versioned directory being the base, and the base directory being the version overlay sounds like typical Microsoft to me.

I guess if I want it to work without fault I need to pull the both tools out, then reinstall starting from 2015.

I'll get around to that in a little and report back.

**Comment by schlamar** [comment]
@Taverius Yes, I think installing Windows SDK 10 Version 10.0.15063.0 (with VS 2017 Build Tools or as separate install) before installing Build Tools 2015 (with Windows SDK Version 10.0.10240.0) will result in this issue.

After installing Build Tools 2015 and then Windows SDK 10 standalone, I had two Windows SDKs installed. In the program manager they are versioned as v10.0.26624 and v10.0.15063.400.

The v10.0.26624 has come with the Build Tools 2015. Uninstalling this SDK caused my build to fail as well.

**Comment by schlamar** [comment]
@fuzzybear3965 pyuv is written in C as well :) Right now, libuv is built with the gyp build system (by calling the vcbuild.bat on Windows) and pyuv is built using the Python build system (setuptools/distutils), linking the compiled libuv library. 

It should be possible to include the libuv sources so they are also built with the Python build system, so the build process should be easier to maintain and it would be more transparent to the user.

**Comment by Taverius** [comment]
Yeah, that satisfied the build system.

I might have also been able to fix it by very carefully ensuring 10.0.10240.0 was fully installed when reinstalling VCBT15, but this actually took less effort, just more time to re-download everything.

Anyway, my original concern here is done, so consider this "solved" for me?

Close at your leisure :)

**Comment by fuzzybear3965** [comment]
@schlamar Oh, okay. I understand a bit better what's going on. Thanks @schlamar.

I wish I could help move the libuv sources into pyuv. I've been looking into distutils' documentations (having had never built a Python package) and it seems that it [still requires a compiler to be installed](https://docs.python.org/2/extending/building.html) external to the installation of Python. So, it doesn't seem possible to avoid a user having to install the Windows C/C++build tools (from VS 2015, VS 2017, Windows SDK, etc.). Is this right? I.e. A fresh Windows + Python installation would still require a C compiler in order to use pyuv (without access to a wheel).

**Comment by schlamar** [comment]
Yes, of course this still requires a VS compiler matching the Python
version.

However, wheels should come soon, so there won't be the need for a compiler
anymore.

Am 31.05.2017 17:27 schrieb "John Rinehart" <notifications@github.com>:

**Comment by fuzzybear3965** [comment]
I understand @schlamar. Since this issue is closed I'll work on compiling this later this week. If I can't get it to work then I'll just wait for the .whl instead of opening up a new issue.

**Comment by schlamar** [comment]
@fuzzybear3965 BTW, you can use the wheels from the CI build of #235 ([x86](https://ci.appveyor.com/project/saghul/pyuv/build/1.0.91/job/0ow590b1nby071sh/artifacts), [x64](https://ci.appveyor.com/project/saghul/pyuv/build/1.0.91/job/hbc52b972sm7ku1f/artifacts)).

Or use Python 3.5 where wheels are already available :)

**Comment by fuzzybear3965** [comment]
Oh, true. I should have checked Appveyor. That insider info, though!

I'm not installing two versions of Python 3, haha. The less clutter on my filesystem, the better.
